<?xml version="1.0" encoding="utf-8"?>

<robot
  name="signaturebot">
  <link name="world"/>

  <link
    name="base_link">
    <inertial>
      <origin
        xyz="0.043978 0.06035 0.0703"
        rpy="0 0 0" />
      <mass
        value="0.3524" />
      <inertia
        ixx="0.0003254"
        ixy="3.9578E-07"
        ixz="2.8883E-06"
        iyy="0.00040409"
        iyz="1.6475E-07"
        izz="0.00040678" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/visual/base_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.75294 0.75294 0.75294 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/collision/base_link.STL" />
      </geometry>
    </collision>
  </link>

  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>

  <link
    name="pitch_link">
    <inertial>
      <origin
        xyz="1.1547E-05 0.0084522 -0.00010354"
        rpy="0 0 0" />
      <mass
        value="0.052828" />
      <inertia
        ixx="3.7725E-06"
        ixy="8.6169E-10"
        ixz="-5.1431E-09"
        iyy="6.3069E-06"
        iyz="4.0016E-11"
        izz="4.2353E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/visual/pitch_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.75294 0.75294 0.75294 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/collision/pitch_link.STL" />
      </geometry>
    </collision>
  </link>

  <joint
    name="pitch"
    type="revolute">
    <origin
      xyz="0.096806 0.060005 0.15025"
      rpy="1.5708 0 1.5708" />
    <parent
      link="base_link" />
    <child
      link="pitch_link" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="-1.571"
      upper="0.698"
      effort="100"
      velocity="1.0" />
    <dynamics
      damping="0.05"
      friction="0.02" />
  </joint>

  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="pitch">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>5</mechanicalReduction>
    </actuator>
  </transmission>

  <link
    name="yaw_link">
    <inertial>
      <origin
        xyz="0.011971 -0.022027 0.040808"
        rpy="0 0 0" />
      <mass
        value="0.15684" />
      <inertia
        ixx="6.4727E-05"
        ixy="8.0682E-06"
        ixz="-8.0047E-07"
        iyy="8.989E-05"
        iyz="5.6188E-07"
        izz="0.00012939" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/visual/yaw_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.49804 0.49804 0.49804 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/collision/yaw_link.STL" />
      </geometry>
    </collision>
  </link>

  <joint
    name="yaw"
    type="revolute">
    <origin
      xyz="0 0 0"
      rpy="-1.5708 0 0" />
    <parent
      link="pitch_link" />
    <child
      link="yaw_link" />
    <axis
      xyz="0 0 -1" />
    <limit
      lower="-0.96"
      upper="1.571"
      effort="100"
      velocity="1.0" />
    <dynamics
      damping="0.05"
      friction="0.02" />
  </joint>

  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="yaw">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>10</mechanicalReduction>
    </actuator>
  </transmission>

  <link
    name="extension_link">
    <inertial>
      <origin
        xyz="0.0061908 -0.065885 9.3556E-05"
        rpy="0 0 0" />
      <mass
        value="0.081074" />
      <inertia
        ixx="5.9014E-05"
        ixy="-3.3423E-08"
        ixz="-2.5282E-11"
        iyy="6.0525E-05"
        iyz="-1.0724E-09"
        izz="4.9084E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/visual/extension_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://fyp/meshes/collision/extension_link.STL" />
      </geometry>
    </collision>
  </link>

  <joint
    name="extension"
    type="prismatic">
    <origin
      xyz="0 -3.6757E-05 -0.050079"
      rpy="-1.5708 0 -1.5708" />
    <parent
      link="yaw_link" />
    <child
      link="extension_link" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="0"
      upper="0.065"
      effort="100"
      velocity="0.5" />
    <dynamics
      damping="0.05"
      friction="0.02" />
  </joint>

  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="extension">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>0.03927</mechanicalReduction>
    </actuator>
  </transmission>

  <gazebo>
    <plugin name="control" filename='libgazebo_ros_control.so'>
      <robotNamespace>/signaturebot</robotNamespace>
    </plugin>
  </gazebo>

  <!--gazebo plugin that publishes JointState messages for all joints, works in conjuctiong with robot_state_publisher to give 3D robot state-->
  <gazebo>
    <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
      <jointName>pitch, yaw, extension</jointName>
      <robotNamespace>/signaturebot</robotNamespace>
      <updateRate>100.0</updateRate>
    </plugin>
  </gazebo>

</robot>
